src/expr/src/linear.rs

```diff
@@ -1356,87 +1357,75 @@ pub mod plan {
+                                    (_, None) => None,
+                                    (x, y) => x.max(y),
+                                };
                             }
                         }
-                    }
-                    Ok(Datum::Null) => {
-                        null_eval = true;
-                    }
-                    x => {
-                        panic!("Non-decimal value in temporal predicate: {:?}", x);
+                        Ok(Datum::Null) => {
+                            null_eval = true;
+                        }
+                        x => {
+                            panic!("Non-decimal value in temporal predicate: {:?}", x);
+                        }
                     }
                 }
             }
-            // At this point, we are "certain" that `lower_bound_i128` is at
-            // least `time`, which means "non-negative" / not needing to be
-            // clamped from below, or its value exceeded `u64::MAX`.
-
-            // Upper bound must be at least lower bound. If we see that the
-            // lower bound is None, we know that it was greater than u64::MAX,
-            // so the upper bound should remain as None because it to would be
-            // forced to be a value greater than `u64::MAX`.
-            if lower_bound.is_some() {
-                // If there are any upper bounds, determine the minimum upper bound.
-                for u in self.upper_bounds.iter() {
+
+            // If the lower bound exceeds `u64::MAX` the update cannot appear in the output.
+            if lower_bound.is_none() {
+                return None.into_iter().chain(None.into_iter());
+            }
+
+            // If there are any upper bounds, determine the minimum upper bound.
+            for u in self.upper_bounds.iter() {
+                // We can cease as soon as the lower and upper bounds match,
+                // as the update will certainly not be produced in that case.
+                if upper_bound != lower_bound {
                     match u.eval(datums, &arena) {
                         Err(e) => {
                             return Some(Err((e, time, diff)))
                                 .into_iter()
                                 .chain(None.into_iter());
                         }
                         Ok(Datum::Numeric(d)) => {
-                            // Upper bound must be at least lower bound, whose
-                            // minimum value is `time`. Negative values will
-                            // always be less than `time`, so will never be
-                            // chosen.
+                            // If the upper bound is negative, it is impossible to satisfy.
+                            // We set the upper bound to the lower bound to indicate this.
                             if d.0.is_negative() {
-                                continue;
+                                upper_bound = lower_bound;
```

Comment:
I was a little confused by the fact that a negative value gets bumped "up" to the lower bound here, but could then be set to another, smaller value in the next iteration. I then saw that line 1450 enforced this invariant again. If the performance of a potential reassignment isn't concerning `upper_bound = Some(0)` would have been a little clearer to me. nbd, though!

Reply:
That's a good point. I was trying to avoid baking in `u64::MIN` anywhere, in case the types changed and e.g. we went with `i64`, this would still be correct. Lemme ponder, as I see what you mean.

Reply:
Probably the right thing (i'll see how gross it looks) it to keep the above, but elsewhere enforce the invariant that `lower_bound <= upper_bound`, rather than having the latter drift down and get corrected later. Then it's just always a thing that goes down, never beyond `lower_bound`.